### PR TITLE
Shuffle metrics 1/4: Add foreground metrics

### LINKED ISSUE #7943: Use metering for P2P shuffling instrumentation
The P2P extensions (primarily the buffers) are implementing their own version for diagnostics, e.g. here https://github.com/dask/distributed/blob/429ef8cc682da36017a28e061d1c773066470fe3/distributed/shuffle/_buffer.py#L267-L272

I think P2P diagnostics would greatly benefit from the metering / fine grained metrics and we should consider replacing (os extending) the custom ctx managers there with the metrics.meter (mind that there is also a special dashboard for P2P shuffling that is using this custom diagnostics. I don't want to break this dashboard).

cc @crusaderky @hendrikmakait

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #7943 

Please read: https://github.com/dask/distributed/issues/7943#issuecomment-1818008425

### CODE REVIEW

File: distributed/shuffle/_core.py [hunk 1]
```diff
@@ -306,7 +308,11 @@ def get_output_partition(
         if not self.transferred:
             raise RuntimeError("`get_output_partition` called before barrier task")
         sync(self._loop, self.flush_receive)
-        return self._get_output_partition(partition_id, key, **kwargs)
+        with (
+            context_meter.meter("shuffle-get-output-noncpu"),
```
Comment (crusaderky) [inline distributed/shuffle/_core.py:312]: I was a bit hesitant to label this "disk", as this also includes GIL contention.

File: distributed/shuffle/_disk.py [hunk 2]
```diff
@@ -179,11 +180,18 @@ def read(self, id: str) -> Any:
...
+                    raise RuntimeError("Already closed")
+                fname = (self.directory / str(id)).resolve()
+                # Note: don't add `with context_meter.meter("shuffle-disk-read"):` to
+                # measure seconds here, as it would shadow "shuffle-get-output-cpu" and
+                # "shuffle-get-output-noncpu". Also, for rechunk it would not measure
+                # the whole disk access, as _read returns memory-mapped buffers.
+                with self.time("read"):
+                    data, size = self._read(fname)
+                context_meter.digest_metric("shuffle-disk-read", 1, "count")
+                context_meter.digest_metric("shuffle-disk-read", size, "bytes")
```
Comment (crusaderky) [inline distributed/shuffle/_disk.py:194]: bytes / count lets you calculate mean read size.

File: distributed/shuffle/_limiter.py [hunk 3]
```diff
@@ -63,19 +65,20 @@ def empty(self) -> bool:
 
     async def wait_for_available(self) -> None:
         """Block until the counter drops below limit"""
+        if not self.full:
+            return
```
Comment (crusaderky) [inline distributed/shuffle/_limiter.py:69]: I've changed the algorithm to exclude non-blocking calls from the mean duration (both with legacy and new metrics).
IMHO it makes more sense like this - but please discuss.

File: distributed/shuffle/_rechunk.py [hunk 4]
```diff
@@ -410,9 +414,15 @@ def _shard_partition(
...
 
+            shards_size += shard.nbytes
+            shards_count += 1
+
             out[self.worker_for[chunk_index]].append(
                 (chunk_index, (shard_index, shard))
             )
+
+        context_meter.digest_metric("shuffle-shards", shards_size, "bytes")
+        context_meter.digest_metric("shuffle-shards", shards_count, "count")
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:425]: bytes / count = mean shard size

File: distributed/shuffle/_worker_plugin.py [hunk 5]
```diff
@@ -277,8 +277,12 @@ def setup(self, worker: Worker) -> None:
...
         self.worker = worker
         self.shuffle_runs = _ShuffleRunManager(self)
-        self.memory_limiter_comms = ResourceLimiter(parse_bytes("100 MiB"))
-        self.memory_limiter_disk = ResourceLimiter(parse_bytes("1 GiB"))
+        self.memory_limiter_comms = ResourceLimiter(
+            parse_bytes("100 MiB"), metrics_label="shuffle-comms-limiter"
+        )
+        self.memory_limiter_disk = ResourceLimiter(
+            parse_bytes("1 GiB"), metrics_label="shuffle-disk-limiter"
+        )
```
Comment (crusaderky) [inline distributed/shuffle/_worker_plugin.py:285]: The metrics emitted by the disk limiter are invisible until #8366

File: distributed/shuffle/_core.py [hunk 6]
```diff
@@ -288,7 +289,8 @@ def add_partition(
         self.raise_if_closed()
         if self.transferred:
             raise RuntimeError(f"Cannot add more partitions to {self}")
-        shards = self._shard_partition(data, partition_id)
+        with context_meter.meter("shuffle-shard-partition", func=thread_time):
```
Comment (crusaderky) [inline distributed/shuffle/_core.py:292]: This cleverly excludes GIL contention. Can't pull the same trick for async methods though.

File: distributed/shuffle/_limiter.py (hunk 3 above)
Comment (hendrikmakait) [reply distributed/shuffle/_limiter.py:69]: I suspect there might be a large performance penalty, but have we considered looking at calculating histograms from (sampled) data points? I'd generally be interested in the distribution of the wait times _including_ non-blocking calls. If that doesn't work, I _think_ I'm more interested in including non-blocking calls here or at least counting them as well. Knowing how long we wait on average (including non-blocking calls) seems to be more relevant to me than knowing how long we wait on average _if we wait_, not knowing anything about how often we wait or don't wait.

File: distributed/shuffle/_rechunk.py (hunk 4 above)
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:426]: Let's look at performance for these in-depth metrics, I'm not sure if collecting shard size here is _always_ useful, so we may want a toggle.

File: distributed/shuffle/_core.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/shuffle/_core.py:292]: For P2P performance, we have identified the GIL as one of the major issues. Would additionally measuring time spent due to GIL contention add significant overhead?

Comment (crusaderky) [reply distributed/shuffle/_core.py:292]: Negligible. Added.

**Comment by crusaderky** [comment]
@hendrikmakait all comments have been addressed

**Review by hendrikmakait — APPROVED** [review]
Thanks, @crusaderky!

**Pull request merged**